pygmt/src/binstats.py

```diff
@@ -87,6 +88,21 @@ def binstats(data, **kwargs):
         Sets the *search_radius* that determines which data points are
         considered close to a node. Append the distance unit.
         Not compatible with ``tiling``.
+    tiling : str
+        **h**\|\ **r**.
+        Instead of circular, possibly overlapping areas, select
+        non-overlapping tiling. Choose between **r**\ ectangular and
+        **h**\ exagonal binning.
+        For **r**, set bin sizes via ``spacing`` and we write
+        the computed statistics to the grid file named in ``outgrid``.
+        For **h**, we write a table with the centers of the hexagons and
+        the computed statistics to standard output (or to the file named
+        in ``outgrid``). Here, the ``spacing`` setting is expected to
```

Comment:
Using `tiling="h"` outputs the statistics to a table, which was why this alias was not added in the orginal implementation, see https://github.com/GenericMappingTools/pygmt/pull/1652#discussion_r881805969.

The way we've handle functions that output to either a grid or table was discussed in https://github.com/GenericMappingTools/pygmt/issues/1536, which is to use a "Two methods in a single Python class" like [`pygmt.triangulate`](https://www.pygmt.org/v0.8.0/api/generated/pygmt.triangulate.html) and [`pygmt.grdhisteq`](https://www.pygmt.org/v0.8.0/api/generated/pygmt.grdhisteq.html). Unfortunately, if we follow this style, that would require breaking changes to `pygmt.binstats` :slightly_smiling_face:

Reply:
Hmm, so @maxrjones mentioned at https://github.com/GenericMappingTools/pygmt/pull/1652#pullrequestreview-1003497323 that the flag to output statistics to a table is actually more similar to https://github.com/GenericMappingTools/pygmt/issues/896? But `gmtbinstats` is not a plotting function, so this is kinda awkward :sweat_smile:

Reply:
Thanks @weiji14 for this clarification and background information. As `tiling` is already used in the [docstring for `search_radius`](https://www.pygmt.org/dev/api/generated/pygmt.binstats.html) I thought that adding the doctstring for `tiling` was simply overlooked...
> search_radius ([float](https://docs.python.org/3/library/functions.html#float) or [str](https://docs.python.org/3/library/stdtypes.html#str)) – Set the search_radius that determines which data points are considered close to a node. Append the distance unit. Not compatible with tiling.

I am unsure how to continue with this PR. Should we leave it open or should we have a separate issue for this?

Reply:
> I am unsure how to continue with this PR. Should we leave it open or should we have a separate issue for this?

Let's leave the PR open for now, the discussion on whether to turn `pygmt.binstats` into two separate methods (e.g. `pygmt.binstats.to_grid` or `pygmt.binstats.to_table`) can happen here. Anything more general can also be discussed on the open thread at https://github.com/GenericMappingTools/pygmt/issues/1536.

Reply:
As I understand this means that it's also not possible to plot the hexagons outlines (polygons )that are used to calculate the stats but to get the center coordinates and stats values for the corresponding hexagon?

Reply:
@michaelgrund what is your concrete use-case here? Please note, that hexagonal binning is only available for Cartesian data and that only the y-increment can be given and GMT calculates the x-increment given the geometry (https://docs.generic-mapping-tools.org/latest/gmtbinstats.html#t). Based on the upstream GMT documentation, I feel it is not possible to get the outline of the hexagons directly.

Are you looking for a figure like this:
|  |  |
| --- | --- |
| ![binstats_hexagons_outline](https://github.com/GenericMappingTools/pygmt/assets/94163266/4646dce4-c73f-47ca-aa55-fd6f93ddf02d) | ![binstats_hexagons_outline_cmap](https://github.com/GenericMappingTools/pygmt/assets/94163266/55122b7a-cf92-4c93-ab0c-c45111eedb40) |

In some cases, one can try to determine the (regular) hexagon based on the center of the hexagon (first two columns of the output file via `outgrid`) and the spacing between the centers in y-direction (value passed to `spacing`). The outline can then be plotted (be careful with width and height of the figure) using the centers of the hexagon and `style="h" + str(determined_diameter_of_hexagon) + "c"`. Additionally, the stats output can be used to color-code the fill of the hexagons (please un-comment the corresponding parts in the code below). However, I feel there are other libraries / packages which are more suitable to generate such a visualization.

_Code to reproduce the figure shown above_:
Test data (Please place it in your work dictionary): [test_data_in.txt](https://github.com/GenericMappingTools/pygmt/files/11635385/test_data_in.txt)

```python
import numpy as np
import pygmt as gmt

# >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
# >>> Adjust for your needs <<<
#
# Give y increment
# Passed to spacing
# In case of hexagon binning only the y increment can be given;
# the x increment is calculated based on the given geometry
# Please see the upsteam GMT documentation
# https://docs.generic-mapping-tools.org/dev/gmtbinstats.html#t
# Last access: 2023/06/01
bin_spacing_y = 1.3

# Give study region
x_min = -1
x_max = 7
y_min = -1
y_max = 5
study_region = [x_min, x_max, y_min, y_max]

# Give size of figure
fig_width = 7

# Give paths and file names
data_in = "test_data_in.txt"
stats_out = "stats_out.txt"
# <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<


#-----------------------------------------------------------------------------
# Be carefulle if you change something here
delta_x = np.abs(x_max - x_min)
delta_y = np.abs(y_max - y_min)

scale_x = fig_width / delta_x

fig_hight = delta_y * scale_x

# Based on the geometry of a regular hexagon
diag_hexagon_temp = ( (bin_spacing_y/2) / np.cos(2*np.pi/360 * 30) ) * 2
diag_hexagon = diag_hexagon_temp * scale_x
bin_spacing_x = diag_hexagon * np.sqrt(3)


#-----------------------------------------------------------------------------
# Bin tabular data based on hexagons
# Calculate stats within hexagons here counts
# Write output to file: center of hexagons and stats
gmt.binstats(
    data=data_in,
    region=study_region,
    spacing=str(bin_spacing_y) + "+e",
    # +e to slightly adjust the max x or y to fit exactly the
    # given increment if needed [Default is to slightly adjust the increment
    # to fit the given domain].
    T="h",  # tiling using hexagons - only for Cartesian data
    statistic="n",  # n counts, a average, m median, z sum
    outgrid=stats_out,
)


#-----------------------------------------------------------------------------
# Plot statistic output
fig = gmt.Figure()

fig.basemap(
    region=study_region,
    projection="X" + str(fig_width) + "c/" + str(fig_hight) + "c",
    frame="a1g1f0.5",
)

# Set up colormap for counts per hexagon
trans_stats = 20
gmt.makecpt(
    cmap="lapaz",
    series=[1, 6, 1],
    reverse=True,
    transparency=trans_stats,
)

# Plot hexagons color-coded by counts
fig.plot(
    data=stats_out,
    style="h" + str(diag_hexagon) + "c",
    pen="0.3p,darkred",
    # cmap=True,    # un-comment for color-coding
    # transparency=trans_stats,  # un-comment for color-coding
)

# Plot centers of hexagons
fig.plot(
    data=stats_out,
    style="p0.15",
    fill="darkred",
)

# Add labels with counts per hexagon
fig.text(
    textfiles=stats_out,
    font="5p",
    offset="0c/-0.2c",
    fill="white@30"
)

# Plot single records
fig.plot(
    data=data_in,
    style="p0.1",
    fill="darkorange",
)

# fig.colorbar(frame="xa1+lcounts per hexagon")  # un-comment for color-coding

fig.show()
# fig.savefig("binstats_hexagons_outline.png")
```